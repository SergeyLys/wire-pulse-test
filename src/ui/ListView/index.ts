export {default as ListView} from './ListView';
export type { ColumnType } from './ListView'