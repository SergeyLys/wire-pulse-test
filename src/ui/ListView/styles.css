.list-container {
  position: relative;
}

.list-header {
  background-color: lightgrey;
  padding: 10px 18px;
  font-weight: bold;
  font-size: 18px;
}

.list-body-item {
  border-bottom: 1px solid lightgray;
  font-size: 16px;
  padding: 10px 15px;
}

.list-body-item:hover {
  background-color: rgb(239, 239, 239);
}

.list-body-col-item {
  padding: 3px;
}

.error-container {
  padding: 30px;
  color: red;
}

.spinner-container {
  padding: 30px;
}

@media (max-width: 480px) {
  .list-header {
    font-size: 16px;
    padding-left: 8px;
    padding-right: 8px;
  }

  .list-body-item {
    font-size: 14px;
    padding-left: 5px;
    padding-right: 5px;
  }
}