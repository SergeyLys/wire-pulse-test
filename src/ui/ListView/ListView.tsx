import { memo } from 'react';
import {Spinner} from '../';
import "./styles.css";

type ColumnType = {
  name: string;
  key: string | number;
  styles: Record<string, any>;
  dataRef: string;
}

type Props<T> = {
  columns: ColumnType[];
  data: T[];
  headerStyles?: Record<string, any>;
  error?: string | null;
  isLoading?: boolean;
}

function ListView<T extends Record<string, any>>({columns, data, headerStyles, error, isLoading}: Props<T>) {
  return (
    <div className="list-container">
      <div className="list-header flex" style={headerStyles}>
        {columns.map((item) => {
          return (
            <div key={`${Number(item.key) * Math.random()}`} style={item.styles}>{item.name}</div>
          )
        })}
      </div>
      <div className="list-body">
        {error && !isLoading && (
          <div className="flex justify-center align-center flex-column error-container">
            Some error occured
            <div>{error}</div>
          </div>
        )}
        {isLoading && (
          <div className='flex justify-center align-center flex-column spinner-container'>
            <Spinner />
          </div>
        )}
        
        {!isLoading && !error && data.length > 0 && data.map((item, index) => {
          return (
            <div key={index * Math.random()} data-testid="listview-item" className="flex list-body-item">{
              columns.map(col => {
                return (
                  <div key={`${Number(col.key) * Math.random()}`} style={{width: col.styles.width}} className="list-body-col-item">{item[col.dataRef]}</div>
                )
              })
            }</div>
          )
        })}
      </div>
    </div>
  )
}

export type {Props, ColumnType};
export default memo(ListView);