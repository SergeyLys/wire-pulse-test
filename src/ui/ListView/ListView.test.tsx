import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import ListView, { Props, ColumnType } from './ListView';

const Component = (props: Props<any>) => (
  <ListView {...props} />
);

const setupComponent = (props?: Record<string, any>) => {
  const columns: ColumnType[] = [
    {
      name: 'Catid',
      styles: {width: 100},
      dataRef: 'catid',
      key: '0',
    },
    {
      name: 'Manufacturer',
      styles: {width: '70%'},
      dataRef: 'manufacturer',
      key: '1',
    },
    {
      name: 'Diameter',
      dataRef: 'diameter',
      key: '2',
      styles: {width: 120},
    }
  ];

  const data = [
    {
      key: 0,
      catid: 213,
      manufacturer: 'Manufactorer 1',
      diameter: `432 mAh`
    },
    {
      key: 1,
      catid: 43,
      manufacturer: 'Manufactorer 3',
      diameter: `432 mAh`
    },
    {
      key: 2,
      catid: 432,
      manufacturer: 'Manufactorer 3',
      diameter: `432 mAh`
    }
  ]

  const defaultProps = {
    columns,
    data,
    error: null,
    isLoading: false
  };

  const mergedProps = {...defaultProps, ...props};

  return <Component {...mergedProps} />;
}

describe('ListView component', () => {
  it('renders correctly', () => {
    const component = setupComponent();

    render(component);

    expect(renderer.create(component).toJSON()).toMatchSnapshot();
    expect(screen.getAllByTestId('listview-item')).toHaveLength(3);
  });

  it('should show loading state when isLoading === true', () => {
    const props = {
      isLoading: true
    }
    const component = setupComponent(props);

    render(component);

    expect(renderer.create(component).toJSON()).toMatchSnapshot();
    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });

  it('should show error state when error is not null', () => {
    const props = {
      error: 'Not found'
    }
    const component = setupComponent(props);
    const tree = renderer.create(component).toJSON();

    render(component);

    expect(tree).toMatchSnapshot();
    expect(screen.getByText('Some error occured')).toBeInTheDocument();
    expect(screen.getByText('Not found')).toBeInTheDocument();
  });
});
