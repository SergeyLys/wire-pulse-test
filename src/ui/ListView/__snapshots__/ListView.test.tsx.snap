// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView component renders correctly 1`] = `
<div
  className="list-container"
>
  <div
    className="list-header flex"
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Catid
    </div>
    <div
      style={
        Object {
          "width": "70%",
        }
      }
    >
      Manufacturer
    </div>
    <div
      style={
        Object {
          "width": 120,
        }
      }
    >
      Diameter
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="flex list-body-item"
      data-testid="listview-item"
    >
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": 100,
          }
        }
      >
        213
      </div>
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": "70%",
          }
        }
      >
        Manufactorer 1
      </div>
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": 120,
          }
        }
      >
        432 mAh
      </div>
    </div>
    <div
      className="flex list-body-item"
      data-testid="listview-item"
    >
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": 100,
          }
        }
      >
        43
      </div>
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": "70%",
          }
        }
      >
        Manufactorer 3
      </div>
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": 120,
          }
        }
      >
        432 mAh
      </div>
    </div>
    <div
      className="flex list-body-item"
      data-testid="listview-item"
    >
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": 100,
          }
        }
      >
        432
      </div>
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": "70%",
          }
        }
      >
        Manufactorer 3
      </div>
      <div
        className="list-body-col-item"
        style={
          Object {
            "width": 120,
          }
        }
      >
        432 mAh
      </div>
    </div>
  </div>
</div>
`;

exports[`ListView component should show error state when error is not null 1`] = `
<div
  className="list-container"
>
  <div
    className="list-header flex"
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Catid
    </div>
    <div
      style={
        Object {
          "width": "70%",
        }
      }
    >
      Manufacturer
    </div>
    <div
      style={
        Object {
          "width": 120,
        }
      }
    >
      Diameter
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="flex justify-center align-center flex-column error-container"
    >
      Some error occured
      <div>
        Not found
      </div>
    </div>
  </div>
</div>
`;

exports[`ListView component should show loading state when isLoading === true 1`] = `
<div
  className="list-container"
>
  <div
    className="list-header flex"
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Catid
    </div>
    <div
      style={
        Object {
          "width": "70%",
        }
      }
    >
      Manufacturer
    </div>
    <div
      style={
        Object {
          "width": 120,
        }
      }
    >
      Diameter
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="flex justify-center align-center flex-column spinner-container"
    >
      <div
        className="circle-loader"
        data-testid="spinner"
      />
    </div>
  </div>
</div>
`;
